ui/components/src/ui-components/tree/controlled/TreeEvents.ts

```diff
@@ -51,6 +52,11 @@ export interface TreeEvents {
    * If Subscription is returned it can be used to stop event handling by calling `unsubscribe()`.
    */
   onCheckboxStateChanged?(event: TreeCheckboxStateChangeEventArgs): Subscription | undefined;
+
+  /**
+   * Raised when rendered nodes in the tree change.
+   */
+  onNodesRendered?: BeUiEvent<TreeRenderedNodeChangeEventArgs>;
```

Comment:
@roluk @grigasp I am not a fan of this approach because it does not right as other type of event are functions. However I could not come up with good solution that would let to get results from this event in [useRenderedNodes](https://github.com/imodeljs/imodeljs/blob/4cdc8cc8505b570a1145bd86074521169e2c5803/presentation/components/src/presentation-components/tree/controlled/TreeHooks.ts#L361) hook. Maybe you have some ideas?

Reply:
The core idea of determining indices of displayed virtualized list items does not fit well with`TreeEventHandler`. I suggest exposing `onItemsRendered` callback on `TreeRendererProps` and save the results in a ref to avoid causing rerenders. The mutable rendered item range ref could then be passed to `useHierarchyAutoUpdate` where it would be accessed by `.current` property since it doesn't need to react to every change to this value.

Reply:
Thanks, I think your approach allows to simplify current implementation quite a lot.

Reply:
Changed